### PR TITLE
Helen ss 136 class based referencing of sides

### PR DESCRIPTION
**Pull request by HelenForbes**

### CODE REVIEW

File: src/cycax/cycad/cycad_side.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+            tuple[float, float, float]: (x, y, z) location of an object.
+        """
+
+        temp_x = self._parent.x_min + sink
+        temp_y = pos[0]
+        temp_z = pos[1]
+
+        return temp_x, temp_y, temp_z
+
+    def _box_size_calc(self, width: float, length: float, depth: float) -> tuple[float, float, float]:
```
Comment (martinslabber) [inline src/cycax/cycad/cycad_side.py:170]: Please also define _box_size_calc in the parent class.

FYI in the future. You only have to comment the method in the parent class (CycaxSide), then you just say "See CycaxSide._box_size_calc for documentation.". Thats only applicable for private methods.

File: src/cycax/cycad/cycad_side.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+        if val is None:
+            return self._parent.x_size
+        else:
+            return val
+
+
+class RightSide(CycaxSide):
+    name = "RIGHT"
+
+    def _box_calculation_bal_s(self, val):
```
Comment (martinslabber) [inline src/cycax/cycad/cycad_side.py:204]: is this used?

File: src/cycax/cycad/cycad_side.py [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+        width: float,
+        depth: float = None,
+        sink: float = 0,
+        center: bool = False,
+    ):
+        """
+        This box will insert a rectangle shape cut out into the object.
+
+        Args:
+            pos (tuple[float, float]): The (x, y) coordinates of the box.
```
Comment (martinslabber) [inline src/cycax/cycad/cycad_side.py:58]: Dont define the type in the docstring, mkdocs will error. It gets the type from the method signature.

**Review by martinslabber — APPROVED** [review]
Some minor comments.

**Comment by HelenForbes** [comment]
I added a save method but forgot to swap branches befor doing so and do not wnat to merge with main incase there are issues with this method.

File: src/cycax/cycad/cuboid.py [hunk 4]
```diff
@@ -25,37 +25,41 @@ def __init__(self, part_no: str, x_size: float, y_size: float, z_size: float, co
...
+        
+class Print3D (Cuboid):
+    
+    """
+    This class will initialize a 3DPrint part at the  location (0,0,0).
+
+    Args:
+        x_size : The size of x.
+        y_size : The size of y.
+        z_size : The siez of z.
```
Comment (martinslabber) [inline src/cycax/cycad/cuboid.py:56]: s/siez/size/

File: src/cycax/cycad/cycad_part.py [hunk 5]
```diff
@@ -256,50 +282,47 @@ def insert_hole(self, hole: Holes):
...
-        Args:
-            x: x location
-            y: y location
-            side: side of object
-
-        Raises:
-            ValueError: if the side is not one of the available and expected sides.
+        This takes the provided part and will create its dictionary and export it to a json
+        """
+        dir_name = f"{os.getcwd()}/{self.part_no}/{self.part_no}"
```
Comment (martinslabber) [inline src/cycax/cycad/cycad_part.py:290]: Probably just one (self.part_no} is needed.

FYI: Good coding practice would be to use os.path.join and not string concatenate with "/" as separators, you might break windows. bleee to bad.

But have a look at [pathlib](https://docs.python.org/3/library/pathlib.html) in Python standard library it is a lot better to use than os. Take a moment and read up on pathlib it is worth the time.
e.g. Below should be the same.
```
from pathlib import Path # Import at the top.
dir_name = Path('.') / self.part_no # Yes they overloaded device to be concat for paths. Pretty cool.
dir_name.mkdir(exist_ok=True)
file_path = dir_name / f"{self.part_no}.json"
json.dump(self.export(), file_path.open("w+"))
```

File: src/cycax/cycad/cycad_part.py [hunk 6]
```diff
@@ -256,50 +282,47 @@ def insert_hole(self, hole: Holes):
...
+            "x": self.x,
+            "y": self.y,
+            "z": self.z,
+            "x_size": self.x_size,
+            "y_size": self.y_size,
+            "z_size": self.z_size,
+            "center": False,
+        }
+
+        dict_part = []
```
Comment (martinslabber) [inline src/cycax/cycad/cycad_part.py:318]: We agreed that export returns a dict (as per the method signature) here it return a list.

File: src/cycax/cycad/cycad_part.py [hunk 7]
```diff
@@ -256,50 +282,47 @@ def insert_hole(self, hole: Holes):
...
-            side: side of object
-
-        Raises:
-            ValueError: if the side is not one of the available and expected sides.
+        This takes the provided part and will create its dictionary and export it to a json
+        """
+        dir_name = f"{os.getcwd()}/{self.part_no}/{self.part_no}"
+        if not os.path.exists(dir_name):
+            os.mkdir(dir_name)
+        with open(f"{dir_name}/{self.part_no}/{self.part_no}.json", "w") as jsonfile:
```
Comment (martinslabber) [inline src/cycax/cycad/cycad_part.py:293]: a {self.part_no} to many.

File: src/cycax/cycad/features.py [hunk 8]
```diff
@@ -145,13 +145,20 @@ class NutCutOut(Location):
 
     """
 
-    nuts = {
-        "type": 3,
-        "diameter": 6.01,
-        "height": 2.4,
+
+    nut_specifications={ #This is a global variable that will be used to cut the nuts by the OpenSCAD engine.
```
Comment (martinslabber) [inline src/cycax/cycad/features.py:149]: Nice.
Probably use this in other Engines beside Openscad.
Is height the same as thickness. It is more common to talk about the thickness of a nut.

File: src/cycax/cycad/cycad_part.py
[the remark this answers is not in the document]
Comment (HelenForbes) [reply src/cycax/cycad/cycad_part.py:1]: This method is used for moving the part around in assembly. I'll make that clear in the function signature.

**Pull request merged**